sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,853 @@
+from ..utils.multiclass import check_classification_targets
+from ..utils.validation import check_is_fitted, check_array, check_X_y
+from ..exceptions import DataDimensionalityWarning
+
+
+class LargeMarginNearestNeighbor(KNeighborsClassifier):
+    """Distance Metric Learning for Large Margin Classification
+
+    Large margin nearest neighbor classification (LMNN) is a machine learning
+    algorithm for metric learning. It learns a (pseudo-)metric in a
+    supervised fashion to improve the classification accuracy of the k-nearest
+    neighbor rule.
+    The main intuition behind LMNN is to learn a pseudometric under which all
+    data instances in the training set are surrounded by at least k instances
+    that share the same class label. If this is achieved, the leave-one-out
+    error is minimized.
+    This implementation follows closely Kilian Weinberger's MATLAB code found
+    at <https://bitbucket.org/mlcircus/lmnn> which solves the unconstrained
+    problem, finding a linear transformation with L-BFGS instead of solving the
+    constrained problem that finds the globally optimal metric.
+
+
+    Parameters
+    ----------
+    L : array, shape (n_features_out, n_features_in), optional, default None
+        An initial linear transformation.
+
+    n_neighbors : int, optional, default 3
+        Number of target neighbors.
+
+    max_iter : int, optional, default 200
+        Maximum number of iterations in the optimization.
+
+    use_pca : bool, optional, default True
+        Whether to use pca to warm-start the linear transformation.
+        If False, the identity will be used.
+
+    tol : float, optional, default 1e-5
+        Convergence tolerance for the optimization.
+
+    n_features_out : int, optional, default None
+        Preferred dimensionality of the inputs after the transformation.
+        If None it is inferred from ``use_pca`` and ``L``.
+
+    max_constraints : int, optional, default 10 million
+        Maximum number of constraints to enforce per iteration.
+
+    use_sparse : bool, optional, default True
+        Whether to use a sparse matrix (default) or a dense matrix for the
+        impostor-pairs storage. Using a sparse matrix, the distance to
+        impostors is computed twice, but it is somewhat faster for larger
+        data sets than using a dense matrix. With a dense matrix, the unique
+        impostor pairs have to be identified explicitly.
+
+    warm_start : bool, optional, default False
+        When set to True, reuse the solution of the previous
+        call to fit as initialization, otherwise, just erase the
+        previous solution.
+
+    max_corrections : int, optional, default 100
+        The maximum number of variable metric corrections
+        used to define the limited memory matrix. (The limited memory BFGS
+        method does not store the full hessian but uses this many terms in an
+        approximation to it.)
+
+    verbose : bool, optional, default False
```

Comment:
On the one hand, I have seen some Estimators have just a boolean `verbose`. 
On the other hand, I had a more complicated `verbose` parameter before but @jnothman commented that scikit-learn usually provides less intricacy in the verbosity options.
So, I'm changing it now to something in between and you can tell me what you think.